pretyping/evarsolve.ml

```diff
@@ -1451,7 +1451,7 @@ let solve_candidates unify flags env evd (evk,argsv) rhs =
         List.map_filter
           (fun c -> filter_compatible_candidates unify flags env evd evi argsv rhs c) l in
       match l' with
-      | [] -> raise IncompatibleCandidates
+      | [] -> raise (IncompatibleCandidates (instantiate_evar_array evi (List.hd l) argsv))
```

Comment:
It might be costly to perform the expansion, you could probably return the existential instead (and tweak the type of the constructors from econstr to existential).

Reply:
More than the evar, the important information is `List.hd l` and the instantiation was actually already computed a few lines above (in `filter_compatible_candidates`). So, the simplest is probably to optimize the computation of `filter_compatible_candidates`, so that we don't compute more than what we really need, if you are ok.

Reply:
I don't really mind, but it's going to be a much more invasive change...